### ISSUE TITLE
Publish bug - nulls in drupal field tables

### ISSUE THREAD

**Issue by dsenalik** [tripal/tripal#1948]
### Tripal Version

4.x

### Drupal Version

10.2, 10.3

### PostgreSQL Version

16

### PHP Version

8.3

### PR #1949 and #1952

PR #1949 will handle sites that already have published content.
PR #1952 will fix publish so that drupal tables have no NULLs and have no extra table rows.

### General Description

I migrated my tripal 3 chado as described here.
https://github.com/tripal/tripal_doc/pull/88
I published organisms, and I am seeing an error.
If I edit, make no changes, and save, then the error disappears for that organism

### Steps to reproduce

To see the error, build a docker on the 4.x branch and do the following
1. Get to your psql prompt and do this
    `INSERT INTO organism (genus, species) VALUES ('Tripalus', 'bogusii');`
2. Publish organism content type
3. View that organism entity
![2024-08-10_4 x](https://github.com/user-attachments/assets/5b977ed3-7b0b-4fa9-9a03-6be98fe59529)

### Error messages

```shell
Deprecated function: str_replace(): Passing null to parameter #3 ($subject) of type array|string is deprecated in Drupal\filter\Element\ProcessedText::preRenderText() (line 101 of core/modules/filter/src/Element/ProcessedText.php).

Drupal\filter\Element\ProcessedText::preRenderText(Array)
call_user_func_array(Array, Array) (Line: 111)
Drupal\Core\Render\Renderer->doTrustedCallback(Array, Array, 'Render #pre_render callbacks must be methods of a class that implements \Drupal\Core\Security\TrustedCallbackInterface or be an anonymous function. The callback was %s. See https://www.drupal.org/node/2966725', 'exception', 'Drupal\Core\Render\Element\RenderCallbackInterface') (Line: 859)
Drupal\Core\Render\Renderer->doCallback('#pre_render', Array, Array) (Line: 421)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 240)
Drupal\Core\Render\Renderer->render(Array) (Line: 475)
Drupal\Core\Template\TwigExtension->escapeFilter(Object, Array, 'html', NULL, 1) (Line: 125)
__TwigTemplate_cafed78bede41823f2a763e1b6c47ebd->doDisplay(Array, Array) (Line: 394)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 367)
Twig\Template->display(Array) (Line: 379)
Twig\Template->render(Array) (Line: 38)
Twig\TemplateWrapper->render(Array) (Line: 39)
twig_render_template('core/themes/olivero/templates/field/field.html.twig', Array) (Line: 348)
Drupal\Core\Theme\ThemeManager->render('field', Array) (Line: 480)
Drupal\Core\Render\Renderer->doRender(Array) (Line: 493)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 240)
Drupal\Core\Render\Renderer->render(Array) (Line: 475)
Drupal\Core\Template\TwigExtension->escapeFilter(Object, Array, 'html', NULL, 1) (Line: 46)
__TwigTemplate_499528482f6b9d60e1aa514303ddea97->doDisplay(Array, Array) (Line: 394)
Twig\Template->displayWithErrorHandling(Array, Array) (Line: 367)
Twig\Template->display(Array) (Line: 379)
Twig\Template->render(Array) (Line: 38)
Twig\TemplateWrapper->render(Array) (Line: 39)
twig_render_template('modules/contrib/tripal/tripal/templates/tripal_entity.html.twig', Array) (Line: 348)
Drupal\Core\Theme\ThemeManager->render('tripal_entity', Array) (Line: 480)
Drupal\Core\Render\Renderer->doRender(Array, ) (Line: 240)
Drupal\Core\Render\Renderer->render(Array, ) (Line: 238)
Drupal\Core\Render\MainContent\HtmlRenderer->Drupal\Core\Render\MainContent\{closure}() (Line: 627)
Drupal\Core\Render\Renderer->executeInRenderContext(Object, Object) (Line: 231)
Drupal\Core\Render\MainContent\HtmlRenderer->prepare(Array, Object, Object) (Line: 128)
Drupal\Core\Render\MainContent\HtmlRenderer->renderResponse(Array, Object, Object) (Line: 90)
Drupal\Core\EventSubscriber\MainContentViewSubscriber->onViewRenderArray(Object, 'kernel.view', Object)
call_user_func(Array, Object, 'kernel.view', Object) (Line: 111)
Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch(Object, 'kernel.view') (Line: 186)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 76)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 58)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 28)
Drupal\Core\StackMiddleware\ContentLength->handle(Object, 1, 1) (Line: 32)
Drupal\big_pipe\StackMiddleware\ContentLength->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 48)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 51)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 36)
Drupal\Core\StackMiddleware\AjaxPageState->handle(Object, 1, 1) (Line: 51)
Drupal\Core\StackMiddleware\StackedHttpKernel->handle(Object, 1, 1) (Line: 704)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```

**Comment by dsenalik** [comment]
This error is being triggered when we pass a NULL when a string is expected. It happens on this line
https://github.com/tripal/tripal/blob/08889235f8f8c07e6b2ab44dc6a74300eb6acec3/tripal/src/Plugin/Field/FieldFormatter/DefaultTripalTextTypeFormatter.php#L44
for the `organism_comment` field.
The `chado.organism` `comment` column is not a null, but it is an empty string.
If I change that line to
`$value_string = $item->get('value')->getValue() ?? '';`
then the error goes away. But that does not answer why it is happening, what is changing when I do an edit and save, it is no longer a NULL

Aha, it is the `tripal_entity__organism_comment` table, there was a NULL in the `organism_comment_value` column. When I edit and save, it changes to an empty string!
Use this command to be able to see the NULLs
`\pset null '<null>'`

~~It's probably happening at this line~~
https://github.com/tripal/tripal/blob/08889235f8f8c07e6b2ab44dc6a74300eb6acec3/tripal/src/Services/TripalPublish.php#L826

The `organism_comment_value` is not part of the sql insert during publish, because it is not part of `$this->required_types`, so it seems to default to NULL on an insert.
```
sql = 
      INSERT INTO {tripal_entity__organism_comment}
        (bundle, deleted, entity_id, revision_id, langcode, delta, organism_comment_record_id) VALUES
(:bundle_2, :deleted_2, :entity_id_2, :revision_id_2, :langcode_2, :delta_2, :organism_comment_record_id_2)

args = array(7) {
  [":bundle_2"]=>
  string(8) "organism"
  [":deleted_2"]=>
  int(0)
  [":entity_id_2"]=>
  string(6) "501920"
  [":revision_id_2"]=>
  int(1)
  [":langcode_2"]=>
  string(3) "und"
  [":delta_2"]=>
  int(0)
  [":organism_comment_record_id_2"]=>
  int(3)
}
```

**Comment by dsenalik** [comment]
I don't like that publish leaves the tables in a slightly different state than manually entering an entity. I wonder if we should be using the drupal entity api instead of directly writing to these tables
https://www.drupal.org/docs/drupal-apis/entity-api/working-with-the-entity-api

But regardless of that, I also feel like we should be filtering out both NULL and empty strings in our formatters, and not adding render arrays for things that aren't there. For example, in `DefaultTripalTextTypeFormatter.php`, we could change
```
      $value_string = $item->get('value')->getValue();
      $elements[$delta] = [
        '#type' => 'processed_text',
        '#text' => $value_string,
        '#format' => $filter_format,
        '#langcode' => $item->getLangcode(),
      ];
```
to
```
      $value_string = $item->get('value')->getValue();
      if (!is_null($value_string) and strlen($value_string)) {
        $elements[$delta] = [
          '#type' => 'processed_text',
          '#text' => $value_string,
          '#format' => $filter_format,
          '#langcode' => $item->getLangcode(),
        ];
      }
```

**Comment by laceysanderson** [comment]
Me + @dsenalik talked about this in weekly codefest and determined the following:

1. We agree that the formatter shouldn't add render arrays when there isn't content. In Tripal 3, the fields knew when they were empty which allowed us to remove the empty field labels... it would be nice to have this again.
2. There is a bug in publish here. We want to ensure that the empty values written to the drupal field tables are of the correct type. I think this would be best done in the StoragePropertyValue object.

More specifically regarding changes to TripalPublish:

As noted by Doug, the following line in TripalPublish is adding a NULL to the SQL since the StoragePropertyValue for that column was not set.

https://github.com/tripal/tripal/blob/08889235f8f8c07e6b2ab44dc6a74300eb6acec3/tripal/src/Services/TripalPublish.php#L826

We want to ensure that StoragePropertyValue::getValue() returns an empty value of the correct type if it was not set. For example, if the property is a string then it should return an empty string, not null.

To do this, we want to 
1. add a method to the \Drupal\tripal\TripalStorage\StoragePropertyValue class that sets the default value.
2. Ensure that when getValue() is called on a StoragePropertyValue whose value is NULL that it instead returns the default value.
3. Add a method to the *StoragePropertyType classes that returns the default value for that property type.
4. Update publish to set the default value of the StoragePropertyValue object when it's created based on the StoragePropertyType object.

The places that need to be changed in publish are:
https://github.com/tripal/tripal/blob/08889235f8f8c07e6b2ab44dc6a74300eb6acec3/tripal/src/Services/TripalPublish.php#L378-L380

https://github.com/tripal/tripal/blob/08889235f8f8c07e6b2ab44dc6a74300eb6acec3/tripal/src/Services/TripalPublish.php#L420-L422

https://github.com/tripal/tripal/blob/08889235f8f8c07e6b2ab44dc6a74300eb6acec3/tripal/src/Services/TripalPublish.php#L430-L432

Where in all these cases we will add a line like this to set the default value for the StoragePropertyValue
```php
$prop_value->setDefaultValue($prop_type->getDefaultValue());
```

**Comment by dsenalik** [comment]
Actually these changes won't do it, the $matches only contain required types. In PR #1952 I have added a function to get the non-required types, and those are handled a bit more simply for the sql in publish and we can get the default value directly without ever needing to create a new StoragePropertyValue

**Comment by dsenalik** [comment]
| 1. We agree that the formatter shouldn't add render arrays when there isn't content. In Tripal 3, the fields knew when they were empty which allowed us to remove the empty field labels... it would be nice to have this again.
PR #1949 fixes this

| 2. There is a bug in publish here. We want to ensure that the empty values written to the drupal field tables are of the correct type. I think this would be best done in the StoragePropertyValue object.
PR #1952 fixes this

**Comment by dsenalik** [comment]
Fixed with merged PR #1949 - but with possible future changes moved to issue #1957
